package/gargoyle/files/www/js/access.js

```diff
@@ -190,50 +253,27 @@ function saveChanges()
 		commands += restartUhttpd ? "killall uhttpd\n/etc/init.d/uhttpd restart\n" : "";
 		//document.getElementById("output").value = commands;
 
-		stopRedirect = false;
-		var param = getParameterDefinition("commands", commands)  + "&" + getParameterDefinition("hash", document.cookie.replace(/^.*hash=/,"").replace(/[\t ;]+.*$/, ""));
+		uciOriginal = uci.clone();
```

Comment:
It concerns me that we are resetting before actually executing the changes. There could still be a fail in the Ajax call yet. Can you justify this one or revert to original call order?

Reply:
I reverted to original call order. Additionally, I used another uciCompare clone from which the old firewall accept sections are removed in order to derive the script commands. In this way, uciOriginal stays original until changes are saved.

But, when I'm not wrong, it makes no difference at the moment. Ajax can indeed fail, but we don't check for success or failure. Every page checks for `req.readyState == 4` but this just means the request is done. `req.state == 200` tells you that the request itself was successful (not the commands which got executed). When I open a page, unplug the Ethernet cable, and then click Save Changes, I get `req.readyState == 4 && req.state != 200`. I think all pages should check for `200`. But for this page, there might be a race condition between the request and restarting uhttpd.

Reply:
Adding error handling for failing Ajax (if at all) affects all pages and can be its own PR. Currently success looks the same like failure, the "Please wait.." dialog just disappears. At least, uciOriginal stays unchanged now and clicking Save Changes again is a valid retry.

Reply:
> At least, uciOriginal stays unchanged now and clicking Save Changes again is a valid retry.

This is the important part.